interfaces/mount/spec.go

```diff
@@ -39,10 +46,134 @@ func (spec *Specification) AddMountEntry(e Entry) error {
 	return nil
 }
 
+// resolveSpecialVariable resolves one of the three $SNAP* variables at the
+// beginning of a given path.  The variables are $SNAP, $SNAP_DATA and
+// $SNAP_COMMON. If there are no variables then $SNAP is implicitly assumed
+// (this is the behavior that was used before the variables were supporter).
+func resolveSpecialVariable(path string, snapInfo *snap.Info) string {
+	if strings.HasPrefix(path, "$SNAP/") || path == "$SNAP" {
+		// NOTE: We use dirs.CoreSnapMountDir here as the path used will be always
+		// inside the mount namespace snap-confine creates and there we will
+		// always have a /snap directory available regardless if the system
+		// we're running on supports this or not.
+		return strings.Replace(path, "$SNAP", filepath.Join(dirs.CoreSnapMountDir, snapInfo.Name(), snapInfo.Revision.String()), 1)
+	}
+	if strings.HasPrefix(path, "$SNAP_DATA/") || path == "$SNAP_DATA" {
+		return strings.Replace(path, "$SNAP_DATA", snapInfo.DataDir(), 1)
+	}
+	if strings.HasPrefix(path, "$SNAP_COMMON/") || path == "$SNAP_COMMON" {
+		return strings.Replace(path, "$SNAP_COMMON", snapInfo.CommonDataDir(), 1)
+	}
+	return path
+}
+
+func mountEntryFromLayout(layout *snap.Layout) (Entry, error) {
+	var entry Entry
+	var nused int
+	if layout.Bind != "" {
+		nused += 1
+	}
+	if layout.Type != "" {
+		nused += 1
+	}
+	if layout.Symlink != "" {
+		nused += 1
+	}
+	if nused != 1 {
+		return entry, fmt.Errorf("layout must define a bind mount, a filesystem mount or a symlink")
```

Comment:
This error is more specific (and IMO more useful) than https://github.com/snapcore/snapd/pull/4505/files#diff-b76be54ad28535ed9b0ac36a2851516eR448 would it be possible to update the one in validation?

Reply:
Yes, I want to unify validation for those two as we'll have bugs lurking otherwise.

Reply:
Done.